File: pathfinding.ts
import { Position, Tile } from './entities';

export interface PathNode {
  x: number;
  y: number;
  g: number; // Cost from start
  h: number; // Heuristic cost to end
  f: number; // Total cost
  parent: PathNode | null;
}

export function findPath(
  start: Position,
  end: Position,
  tiles: Tile[],
  width: number,
  height: number,
  maxDistance: number = 15
): Position[] {
  // Simple A* pathfinding implementation
  const openList: PathNode[] = [];
  const closedList: Set<string> = new Set();
  
  const startNode: PathNode = {
    x: start.x,
    y: start.y,
    g: 0,
    h: heuristic(start, end),
    f: 0,
    parent: null
  };
  startNode.f = startNode.g + startNode.h;
  
  openList.push(startNode);
  
  while (openList.length > 0) {
    // Find node with lowest f score
    let currentIndex = 0;
    for (let i = 1; i < openList.length; i++) {
      if (openList[i].f < openList[currentIndex].f) {
        currentIndex = i;
      }
    }
    
    const currentNode = openList[currentIndex];
    openList.splice(currentIndex, 1);
    closedList.add(`${currentNode.x},${currentNode.y}`);
    
    // Check if we reached the target
    if (currentNode.x === end.x && currentNode.y === end.y) {
      const path: Position[] = [];
      let node: PathNode | null = currentNode;
      while (node) {
        path.unshift({ x: node.x, y: node.y });
        node = node.parent;
      }
      return path.slice(1); // Remove the starting position
    }
    
    // Check neighbors
    const neighbors = [
      { x: currentNode.x + 1, y: currentNode.y },
      { x: currentNode.x - 1, y: currentNode.y },
      { x: currentNode.x, y: currentNode.y + 1 },
      { x: currentNode.x, y: currentNode.y - 1 }
    ];
    
    for (const neighbor of neighbors) {
      const key = `${neighbor.x},${neighbor.y}`;
      
      // Check bounds
      if (neighbor.x < 0 || neighbor.x >= width || neighbor.y < 0 || neighbor.y >= height) {
        continue;
      }
      
      // Check if already processed
      if (closedList.has(key)) {
        continue;
      }
      
      // Check if walkable
      const tileIndex = neighbor.y * width + neighbor.x;
      const tile = tiles[tileIndex];
      if (tile.type === 'wall') {
        continue;
      }
      
      // Check distance limit
      if (heuristic(start, neighbor) > maxDistance) {
        continue;
      }
      
      const g = currentNode.g + 1;
      const h = heuristic(neighbor, end);
      const f = g + h;
      
      // Check if this path to neighbor is better
      const existingNodeIndex = openList.findIndex(node => node.x === neighbor.x && node.y === neighbor.y);
      if (existingNodeIndex >= 0) {
        if (g < openList[existingNodeIndex].g) {
          openList[existingNodeIndex].g = g;
          openList[existingNodeIndex].f = f;
          openList[existingNodeIndex].parent = currentNode;
        }
      } else {
        const neighborNode: PathNode = {
          x: neighbor.x,
          y: neighbor.y,
          g,
          h,
          f,
          parent: currentNode
        };
        openList.push(neighborNode);
      }
    }
  }
  
  return []; // No path found
}

function heuristic(a: Position, b: Position): number {
  return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
}

export function getSimpleDirection(from: Position, to: Position): Position {
  const dx = to.x - from.x;
  const dy = to.y - from.y;
  
  // Normalize to -1, 0, or 1
  const moveX = dx === 0 ? 0 : dx > 0 ? 1 : -1;
  const moveY = dy === 0 ? 0 : dy > 0 ? 1 : -1;
  
  // Prefer moving in the direction of the larger difference
  if (Math.abs(dx) > Math.abs(dy)) {
    return { x: moveX, y: 0 };
  } else {
    return { x: 0, y: moveY };
  }
}

File: python_game/game/pathfinding.py
from dataclasses import dataclass
from typing import List, Optional, Set
from game.game_objects import Position, Tile

@dataclass
class PathNode:
    x: int
    y: int
    g: int
    h: int
    f: int
    parent: Optional['PathNode'] = None

def find_path(start: Position, end: Position, tiles: List[Tile], width: int, height: int, max_distance: int = 15) -> List[Position]:
    open_list: List[PathNode] = []
    closed_list: Set[str] = set()
    
    start_node = PathNode(
        x=start.x,
        y=start.y,
        g=0,
        h=heuristic(start, end),
        f=0
    )
    start_node.f = start_node.g + start_node.h
    
    open_list.append(start_node)
    
    while open_list:
        current_index = 0
        for i in range(1, len(open_list)):
            if open_list[i].f < open_list[current_index].f:
                current_index = i
        
        current_node = open_list.pop(current_index)
        closed_list.add(f"{current_node.x},{current_node.y}")
        
        if current_node.x == end.x and current_node.y == end.y:
            path = []
            node = current_node
            while node:
                path.insert(0, Position(x=node.x, y=node.y))
                node = node.parent
            return path[1:]
        
        neighbors = [
            {'x': current_node.x + 1, 'y': current_node.y},
            {'x': current_node.x - 1, 'y': current_node.y},
            {'x': current_node.x, 'y': current_node.y + 1},
            {'x': current_node.x, 'y': current_node.y - 1}
        ]
        
        for neighbor in neighbors:
            key = f"{neighbor['x']},{neighbor['y']}"
            
            if neighbor['x'] < 0 or neighbor['x'] >= width or neighbor['y'] < 0 or neighbor['y'] >= height:
                continue
            
            if key in closed_list:
                continue
            
            tile_index = neighbor['y'] * width + neighbor['x']
            tile = tiles[tile_index]
            if tile.type == 'wall':
                continue
            
            if heuristic(Position(x=start.x, y=start.y), Position(x=neighbor['x'], y=neighbor['y'])) > max_distance:
                continue
            
            g = current_node.g + 1
            h = heuristic(Position(x=neighbor['x'], y=neighbor['y']), end)
            f = g + h
            
            existing_node_index = None
            for i, node in enumerate(open_list):
                if node.x == neighbor['x'] and node.y == neighbor['y']:
                    existing_node_index = i
                    break
            
            if existing_node_index is not None:
                if g < open_list[existing_node_index].g:
                    open_list[existing_node_index].g = g
                    open_list[existing_node_index].f = f
                    open_list[existing_node_index].parent = current_node
            else:
                neighbor_node = PathNode(
                    x=neighbor['x'],
                    y=neighbor['y'],
                    g=g,
                    h=h,
                    f=f,
                    parent=current_node
                )
                open_list.append(neighbor_node)
    
    return []

def heuristic(a: Position, b: Position) -> int:
    return abs(a.x - b.x) + abs(a.y - b.y)

def get_simple_direction(from_pos: Position, to_pos: Position) -> Position:
    dx = to_pos.x - from_pos.x
    dy = to_pos.y - from_pos.y
    
    move_x = 0 if dx == 0 else (1 if dx > 0 else -1)
    move_y = 0 if dy == 0 else (1 if dy > 0 else -1)
    
    if abs(dx) > abs(dy):
        return Position(x=move_x, y=0)
    else:
        return Position(x=0, y=move_y)
